### PR TITLE
Ext cross corr

### PR DESCRIPTION
**Pull request by elvinpoole**
Add stage to compute two-point between external healsparse maps (e.g. survey properties or dust maps) and TXPipe tracers: Galaxy density and Shear

### CODE REVIEW

File: txpipe/twopoint.py [hunk 1]
```diff
@@ -250,12 +253,14 @@ def add_data_points(self, S, results):
         GAMMAT = sacc.standard_types.galaxy_shearDensity_xi_t
         GAMMAX = sacc.standard_types.galaxy_shearDensity_xi_x
         WTHETA = sacc.standard_types.galaxy_density_xi
+        POS_EXT = "pos_ext"
```
Comment (empEvil) [inline txpipe/twopoint.py:256]: Why not buiild a sacc data type for it? you can do something like `GAMMATS = sacc.build_data_type_name(
            "galaxy", ["shear", "Density"], "xi", subtype="ts"
        )` (Example pulled from the self-calibration extension

File: txpipe/twopoint.py [hunk 2]
```diff
@@ -250,12 +253,14 @@ def add_data_points(self, S, results):
...
+        SHEAR_EXT = "shear_ext"
 
         comb = []
         for index, d in enumerate(results):
             # First the tracers and generic tags
-            tracer1 = f"source_{d.i}" if d.corr_type in [XI, GAMMAT] else f"lens_{d.i}"
-            tracer2 = f"source_{d.j}" if d.corr_type in [XI] else f"lens_{d.j}"
+            tracer1 = f"source_{d.i}" if d.corr_type in [XI, GAMMAT, SHEAR_EXT] else f"lens_{d.i}"
+            tracer2 = f"source_{d.j}" if d.corr_type in [XI] else (f"external_{d.j}" if d.corr_type in [POS_EXT, SHEAR_EXT] else f"lens_{d.j}")
```
Comment (empEvil) [inline txpipe/twopoint.py:263]: does this way of writing an `elif` actually work?

File: txpipe/twopoint.py [hunk 3]
```diff
@@ -283,6 +288,7 @@ def add_data_points(self, S, results):
                         error=xiperr[i],
                         npair=npair[i],
                         weight=weight[i],
+                        tracers_later=tracers_later,
```
Comment (empEvil) [inline txpipe/twopoint.py:291]: what does this thing do? (if answered later disregard my message)

File: txpipe/twopoint.py [hunk 4]
```diff
@@ -493,6 +503,18 @@ def call_treecorr(self, i, j, k):
         elif k == POS_POS:
             xx = self.calculate_pos_pos(i, j)
             xtype = sacc.standard_types.galaxy_density_xi
+        elif k == POS_EXT:
```
Comment (empEvil) [inline txpipe/twopoint.py:506]: Does this not cause issues if someone tries to run just TXTwopoint on an external, but not the class? i.e. Maybe we ought to move the entire function `call_treecorr` to the extension class

File: txpipe/twopoint.py [hunk 5]
```diff
@@ -1092,6 +1114,235 @@ def calculate_pos_pos(self, i, j):
...
+            w=mask[pix], #weight pixels by their fracdet
+            k=map_d,
+            ra_units="degree",
+            dec_units="degree",
+            patch_centers=self.get_input("patch_centers"),
+        )
+        return cat
+
+
+    def write_output(self, source_list, lens_list, meta, results):
```
Comment (empEvil) [inline txpipe/twopoint.py:1291]: I think you need to provide something more to SACC to get it to actually save your file no?

**Review by empEvil — COMMENTED** [review]
Overall looks good, but I am still trying to understand how you can get away with not defining the correlation type with SACC

File: txpipe/twopoint.py (hunk 3 above)
Comment (elvinpoole) [reply txpipe/twopoint.py:306]: One of the tracers is the external map. This isn't a known tracer in sacc so it gives an error unless you set tracers_later=True

File: txpipe/twopoint.py (hunk 5 above)
Comment (elvinpoole) [reply txpipe/twopoint.py:1302]: seems to be saving for me. I'll add the example to demonstrate

File: txpipe/twopoint.py (hunk 1 above)
Comment (elvinpoole) [reply txpipe/twopoint.py]: I didn't know you could do this! updated it

File: txpipe/twopoint.py (hunk 2 above)
Comment (elvinpoole) [reply txpipe/twopoint.py]: i think it worked, but probably wasn't very clear. I made it a longer set of if and elif statements

File: txpipe/twopoint.py (hunk 5 above)
Comment (elvinpoole) [reply txpipe/twopoint.py:1302]: example added

File: txpipe/twopoint.py (hunk 4 above)
Comment (empEvil) [reply txpipe/twopoint.py:521]: What do you think, should we add it to the subclass, or do you want to keep it here? might be worth to get @joezuntz  input once he is back?

**Review by empEvil — COMMENTED** [review]
Think we just need to decide on what should be in the subclass and the main class. Other than that I think we are good

File: txpipe/twopoint.py (hunk 4 above)
Comment (elvinpoole) [reply txpipe/twopoint.py]: ok sounds good. I moved it

**Comment by elvinpoole** [comment]
removed all the edits to the base class and moved them to the new external cross correlation class

**Pull request merged**